src/peft/tuners/lora/eva.py

```diff
@@ -0,0 +1,673 @@
+    @torch.no_grad()
+    def __call__(self, model, input, output):
+        previous_components = None
+        if hasattr(self.svd, "components_"):
+            previous_components = self.svd.components_.clone().detach()
+        states = self.prepare_layer_inputs(input)
+        # check if batch sizes is more than the number of components
+        if states.size(0) < self.n_components:
+            print(f"skipping SVD for {self.name} because there are less than {self.n_components} examples")
+            return
+        self.svd.partial_fit(states.to(torch.float32))
+        # add if statement to check if we are in the first step where previous_components is None
+        if previous_components is None:
+            return
+        components = self.svd.components_
+        if len(components.shape) == 1:
+            components = components.reshape(1, -1)
+            previous_components = previous_components.reshape(1, -1)
+        # consider as converged if enough components have converged via cossim
+        sim = torch.nn.functional.cosine_similarity(components, previous_components)
+        self.converged = sim >= self.sim_thresh
+
+
+# This is used to determine if inputs of two different layers are equal. For such cases, SVD
+# needs to be done for only for one of the equal inputs.
+class HashHook(_Hook):
+    """
+    A forward hook for hashing layer inputs. The hook is designed to be registered to a PyTorch module using the
+    `register_forward_hook` method.
+
+    This hook hashes the inputs of a specified layer during the forward pass of a neural network and stores the hash
+    values for later analysis or comparison.
+
+    Args:
+        name (str): Name of the layer to which this hook is attached. hashed_inputs (list): List of hashed inputs.
+        prepare_layer_inputs_fn (Optional[callable]): Function to prepare layer inputs for hashing.
+    """
+
+    def __init__(
+        self,
+        name: str,
+        prepare_layer_inputs_fn: Optional[callable] = None,
+    ):
+        super().__init__(name, prepare_layer_inputs_fn)
+        self.hashed_inputs = []
+
+    @staticmethod
+    def hash_fn(tensor):
+        return hash(tuple(tensor.view(-1).tolist()))
+
+    @torch.no_grad()
+    def __call__(self, model, input, output):
+        x = self.prepare_layer_inputs(input).cpu()
+        self.hashed_inputs.append(self.hash_fn(x))
+
+
+def find_equal_values(dictionary: dict) -> dict:
+    """
+    Find keys in a dictionary that have the same value.
+
+    This function takes a dictionary and returns a new dictionary containing keys that have the same value. The keys in
+    the output dictionary are the values from the input dictionary, and the values are lists of keys that share the
+    same value.
+    """
+    value_dict = defaultdict(list)
+    for k, v in dictionary.items():
+        value_dict[v].append(k)
+    return {k: v for k, v in value_dict.items() if len(v) > 1}
+
+
+def get_target_name_key(name, pattern_keys):
```

Comment:
This can be replaced by `get_pattern_key` if #2195 is merged first, right? Let's wait with this PR for the other one then.

Reply:
yes correct!